crates/core/src/host/v8/de.rs

```diff
@@ -0,0 +1,420 @@
+}
+
+impl<'s> DeserializerCommon<'_, 's> {
+    fn reborrow(&mut self) -> DeserializerCommon<'_, 's> {
+        DeserializerCommon {
+            scope: self.scope,
+            key_cache: self.key_cache,
+        }
+    }
+}
+
+/// The possible errors that [`Deserializer`] can produce.
+#[derive(From)]
+pub(super) enum Error<'s> {
+    Value(Local<'s, Value>),
+    Exception(ExceptionThrown),
+    Custom(String),
+}
+
+impl de::Error for Error<'_> {
+    fn custom(msg: impl fmt::Display) -> Self {
+        Self::Custom(msg.to_string())
+    }
+}
+
+/// Returns a scratch buffer to fill when deserializing strings.
+fn scratch_buf<const N: usize>() -> [MaybeUninit<u8>; N] {
+    [const { MaybeUninit::uninit() }; N]
+}
+
+/// A cache for frequently used strings to avoid re-interning them.
+#[derive(Default)]
+pub(super) struct KeyCache {
+    /// The `tag` property for sum values in JS.
+    tag: Option<Global<v8::String>>,
+    /// The `value` property for sum values in JS.
+    value: Option<Global<v8::String>>,
+}
+
+impl KeyCache {
+    /// Returns the `tag` property name.
+    pub(super) fn tag<'s>(&mut self, scope: &mut HandleScope<'s>) -> Local<'s, v8::String> {
+        Self::get_or_create_key(scope, &mut self.tag, "tag")
+    }
+
+    /// Returns the `value` property name.
+    pub(super) fn value<'s>(&mut self, scope: &mut HandleScope<'s>) -> Local<'s, v8::String> {
+        Self::get_or_create_key(scope, &mut self.value, "value")
+    }
+
+    /// Returns an interned string corresponding to `string`
+    /// and memoizes the creation on the v8 side.
+    fn get_or_create_key<'s>(
+        scope: &mut HandleScope<'s>,
+        slot: &mut Option<Global<v8::String>>,
+        string: &str,
+    ) -> Local<'s, v8::String> {
+        if let Some(s) = &*slot {
+            v8::Local::new(scope, s)
+        } else {
+            let s = v8_interned_string(scope, string);
+            *slot = Some(v8::Global::new(scope, s));
+            s
+        }
+    }
+}
+
+// Creates an interned [`v8::String`].
+fn v8_interned_string<'s>(scope: &mut HandleScope<'s>, field: &str) -> Local<'s, v8::String> {
+    // Internalized v8 strings are significantly faster than "normal" v8 strings
+    // since v8 deduplicates re-used strings minimizing new allocations
+    // see: https://github.com/v8/v8/blob/14ac92e02cc3db38131a57e75e2392529f405f2f/include/v8.h#L3165-L3171
+    v8::String::new_from_utf8(scope, field.as_ref(), v8::NewStringType::Internalized).unwrap()
+}
+
+/// Extracts a reference `&'s T` from an owned V8 [`Local<'s, T>`].
+///
+/// The lifetime `'s` is that of the [`HandleScope<'s>`].
+/// This ensures that the reference to `T` won't outlive the `HandleScope`.
+fn deref_local<'s, T>(local: Local<'s, T>) -> &'s T {
+    let reference = local.borrow();
+    // SAFETY: Lifetime extend `'0` to `'s`.
+    // This is safe as the returned reference `&'s T`
+    // will not outlive its `HandleScope<'s, _>`,
+    // as both are tied to the lifetime `'s`.
```

Comment:
This comment on its own is not sufficient to convince me. I was unsure whether the `Local` value itself was keeping its referent alive in the V8 GC. It turns out that the enclosing `HandleScope` does that, and the `Local` is just a normal reference within that scope. I had to go check the docs for `Local`, where I found the following paragraph:

> It is safe to extract the object stored in the handle by dereferencing the handle (for instance, to extract the *Object from a Local<Object>); the value will still be governed by a handle behind the scenes and the same rules apply to these values as to their handles.

I recommend linking to `Local` and also either quoting or paraphrasing that point.